signer/migrations/0011__materialized_chains.sql

```diff
@@ -0,0 +1,148 @@
+-- to identify the specific chain tip for which the canonical chain is
+-- materialized. If no Stacks block(s) exist for a given Bitcoin block, the
+-- `stacks_block_hash` and `stacks_block_height` columns will be `NULL` for that
+-- Bitcoin block hash.
+CREATE TABLE sbtc_signer.canonical_chains (
+    run_id INT,
+    bitcoin_chain_tip BYTEA,
+    bitcoin_block_hash BYTEA,
+    bitcoin_block_height BIGINT,
+    stacks_block_hash BYTEA,
+    stacks_block_height BIGINT
+);
+
+-- An intermediate view to `canonical_chains` which can be used in queries to
+-- avoid needing to change query syntax if the underlying data source changes.
+-- Note that this view will return multiple canonical chains, so you need to
+-- filter the results to the specific chain tip which you are interested in.
+CREATE VIEW sbtc_signer.canonical_chains_view AS
+SELECT
+    bitcoin_chain_tip,
+    bitcoin_block_hash,
+    bitcoin_block_height,
+    stacks_block_hash,
+    stacks_block_height
+FROM sbtc_signer.canonical_chains;
+
+-- A helper view to get only the canonical Bitcoin chain information from
+-- `canonical_chains` if you're not interested in Stacks blocks. Note that if
+-- used in queries, the `bitcoin_chain_tip` column should be used to filter the
+-- results to the specific chain tip which you are interested in. The results
+-- are not ordered by block height, so you should do that yourself if you
+-- require the results to be ordered.
+CREATE VIEW sbtc_signer.canonical_bitcoin_chain_view AS
+SELECT DISTINCT
+    bitcoin_chain_tip,
+    bitcoin_block_hash,
+    bitcoin_block_height
+FROM sbtc_signer.canonical_chains;
+
+-- Indexes to support common queries on the canonical chains table.
+CREATE INDEX ix_canonical_chains_run_id ON sbtc_signer.canonical_chains(run_id);
+CREATE INDEX ix_canonical_chains_bitcoin_chain_tip ON sbtc_signer.canonical_chains(bitcoin_chain_tip);
+CREATE INDEX ix_canonical_chains_bitcoin_block_hash ON sbtc_signer.canonical_chains(bitcoin_chain_tip, bitcoin_block_hash);
+CREATE INDEX ix_canonical_chains_bitcoin_block_height ON sbtc_signer.canonical_chains(bitcoin_chain_tip, bitcoin_block_height);
+CREATE INDEX ix_canonical_chains_stacks_block_hash ON sbtc_signer.canonical_chains(bitcoin_chain_tip, stacks_block_hash);
+CREATE INDEX ix_canonical_chains_stacks_block_height ON sbtc_signer.canonical_chains(bitcoin_chain_tip, stacks_block_height);
+
+-- New sequence which will be used to generate the run_id for each materialized
+-- view canonical chain.
+CREATE SEQUENCE sbtc_signer.canonical_chains_run_id_seq;
+
+-- Function to materialize the canonical chains for a given bitcoin chain tip
+-- and the maximum depth of the bitcoin blockchain to consider.
+--
+-- This function returns the number of rows written to the `canonical_chains`
+-- table upon success, and `-1` if rows already exist for the given chain tip.
+CREATE OR REPLACE FUNCTION sbtc_signer.materialize_canonical_chains(chain_tip BYTEA, max_depth INT)
```

Comment:
Do we need to lock/run this whole function in a transaction? Because it's fine for the coordinator, but the signers may start to query this for a bitcoin tip that is still being worked on.

Reply:
postgres runs all statements within a function in a transaction (i.e. everything between begin/end), so it should be fine. [Link](https://cooxe.dev/2023-10-06-transaction-in-postgresql-function.html)

Reply:
Oh, and the default isolation level is `READ COMMITTED`, so a reader will never get partially-written/uncommitted data.